kitchen/bin/waiter-init

```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+#
+# A wrapper script for the Waiter-specific setup for a user command in a Waiter-K8s pod.
+# The script is usually invoked by prepending it to the user's Waiter command.
+#
+# A single argument is expected, the user's command string,
+# which is executed as its own bash shell process.
+
+# Ignore SIGTERM sent by Kubernetes on pod deletion,
+# waiting for SIGKILL (force delete) before exiting.
+trap : SIGTERM
+
+# Run the user's Waiter app command,
+# copying stdout and stderr to respectively named files.
+# We tee the output so that stdout and stderr are still accessible
+# via the Kubernetes `kubectl logs <pod-name>` command.
+/bin/bash -c "$1" 1> >(tee stdout) 2> >(tee stderr 1>&2)
+exit_code=$?
+
+# If the sub-command exited 143 (usually indicated it was killed via SIGTERM),
+# then sleep indefinitely, awaiting a SIGKILL from Kubernetes
```

Comment:
Should we sleep for some definite period instead of indefinitely?

Reply:
No, I don't see any advantage to picking an arbitrary time period instead of just letting it sleep forever. In all cases, Kubernetes should send a SIGKILL to the pod's containers long before our arbitrary timeout expires, meaning that whatever limit we choose would be essentially indefinite anyway. The downside is that if we chose an arbitrary timeout that's too short, we might accidentally change something in the Waiter-K8s implementation that causes a race. It seems better to me to let this sleep forever.

Reply:
I'd feel more at ease if we slept for a long enough period, e.g. 1 hour, that avoids concerns about the data race you mention above. Can you be more specific as to what kind of change would introduce a data race for an exited command?

Reply:
If we set the sleep period to 15 minutes, but later change our default pod termination grace period in our ReplicaSet spec to 20 minutes, then the pod might die (due to this sleep ending) before we're expecting it to. However, if the other cleanup is taking so long that the grace-period timeout is kicking in, we're in a bad situation anyway... So I guess that's not really a concern after all.

I can't think of a benefit to picking an arbitrary cutoff here. If the pod was sigtermed by Kuberentes, then it's going to get sigkilled after the grace period expires. If the sigterm was part of a Waiter-K8s scale-down operation, then Waiter will do a force-delete (sigkill) as soon as it updates the replica count (hopefully < 1 second delay). In all other cases, Kubernetes will detect that the pod is unhealthy (health checks failing) and restart it.

Reply:
I changed the delay to 15 minutes. That seems sufficiently long, but it's still arbitrary.